### ISSUE TITLE
this used to be my blog — Graphing when your Facebook friends are awake

### ISSUE THREAD

**Issue by tg-z** [tg-z/web-clip#390]
_This blog has moved! This post and other mistakes are now at [https://mango.pdf.zone](https://t.umblr.com/redirect?z=https%3A%2F%2Fmango.pdf.zone&t=NmE5Yjg4ZjIwMzNiOGQ5OTY3ODVmMTA5MDdjYTQ2YzYwYzQ1Y2RlMCxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0)_

* * *

Look I’m not really sure why but I think I made a thing that makes graphs of when people are online on [Facebook](https://t.umblr.com/redirect?z=http%3A%2F%2Fgabegaming.com&t=MDhhYzFiZWNhMTc4YjZmMWIzNTUwZmM4MTU3OTRjNjcwODA0NWEzZixZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0). It sounds kinda creepy and uh it is. Read along so you, too, can be the NSA. ˙ ͜ʟ˙

* * *

![](https://media.tumblr.com/1de7e3691413a7ff657d151193adc508/tumblr_inline_o2iuleaQX21t2msv6_500.png)

* * *

Little green dots
-----------------

You know those green dots on the sidebar on [Facebook](https://t.umblr.com/redirect?z=http%3A%2F%2Fgabegaming.com&t=MDhhYzFiZWNhMTc4YjZmMWIzNTUwZmM4MTU3OTRjNjcwODA0NWEzZixZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) that tell you who’s online? How do they get there?  Also there are times next to people who are offline. What are those about?

![](https://media.tumblr.com/12ef3a8c87e9dd14268524da817e8f9d/tumblr_inline_o2gnqkRZLc1t2msv6_400.png)

I was wondering the same things, and so one day I decided to 360 noscope hack [Facebook](https://t.umblr.com/redirect?z=http%3A%2F%2Fniceme.me&t=ODUzZWJiNTg3ZDY1YWYyNTA1ODNhNmRhNDNiZmJmMzQwZTFlMzNmYyxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) by right clicking and selecting “Inspect Element”.

![](https://media.tumblr.com/b7705f1bc7762f6a84bec166b4883fcc/tumblr_inline_o2l7c3GpWr1t2msv6_500.png)

_I’M **IN**_
------------

We did it team. Anyway alright uhhhh let’s just uh snoop around here _reallllll sneaky like_

![](https://media.tumblr.com/78ab3974532f3f32b7cc51a1082f4f05/tumblr_inline_o2l7dtrAS81t2msv6_400.png)

If you reload the page you’ll see approximately fifty-bajillion network requests go off as [Facebook](https://t.umblr.com/redirect?z=http%3A%2F%2Fwww.thinkb4uclick.ie%2F&t=ZTlhYjA3ZjZlMzQ3ZjhmYWE0MzA3NWVhM2U5NWVlOGU2ZDlmZWZhNyxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) desperately tries to load all the junk that it needs to display [facebook.com](https://t.umblr.com/redirect?z=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3DdQw4w9WgXcQ&t=OTNiZmY5NzNhNGU5YTAwM2IwM2IzOWRiZGQ1ODY2YzFkZWI3ZmVmZSxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0). 

You might be wondering at this point why I decided to look for interesting things in this mess instead of, I dunno, getting out more, getting a cat, that sorta thing. Anyway hey look a heading

Finding the good stuff
----------------------

![](https://media.tumblr.com/d9dfc86de408f5cb8d96f08ed1052504/tumblr_inline_o2l7ezOeS11t2msv6_500.png)

What’s this “pull” thing?

![](https://media.tumblr.com/f7f0877a8d116b07861116bbdef8c424/tumblr_inline_o2l7gaDn8L1t2msv6_500.png)

THAT looks like some #datascience right there. This is the kind of **100% legit secret undocumented “API”** that we came here for. Let’s do some reverse-engineering.

It looks like a mapping of [Facebook](https://t.umblr.com/redirect?z=http%3A%2F%2Fgabegaming.com&t=MDhhYzFiZWNhMTc4YjZmMWIzNTUwZmM4MTU3OTRjNjcwODA0NWEzZixZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) user ids to… their online status? But there’s more than one value? “webStatus” and “fbAppStatus” are both there. What’s more, it tells you what the person is _doing_ on each of the different kinds of statuses. 

For example:

*   “messengerStatus” : “invisible” means they’re not online on the Facebook Messenger app.   
    
*   “webStatus”: “idle” means their web browser is logged in to [Facebook](https://t.umblr.com/redirect?z=http%3A%2F%2Foneu.se&t=YTI0NWYwZGQ2M2ZhN2M1N2E4ZDlhZTA3YjE2Yjc3MmZhODBmNWMxYSxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0), and has the page open, but they aren’t doing anything on the site like moving their mouse or talking to anyone.  
    
*   Since we have both of these _at the same time_, we can tell that this person is likely not using their phone, and that they were using [facebook.com](https://t.umblr.com/redirect?z=http%3A%2F%2Fwww.logobird.com%2Fwp-content%2Fuploads%2F2011%2F04%2Ffacebook-procrastination.jpg&t=YzA4YTk1YmIyZjIxZDA0YjlmOTYzMTUyMGVmZjkyZmM2ZmQ1MzUwZixZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) recently, but not right now.

That’s already a little creepy that we can tell that about people. But can we do more with this?

You might also notice that there is a value called “la” that is a big integer that starts with “14″. If you I dunno, didn’t have a lot of friends in high school, you might recognise that as a [UNIX time stamp](https://t.umblr.com/redirect?z=https%3A%2F%2Fen.wikipedia.org%2Fwiki%2FUnix_time&t=Yjg5NjYyYjQwYmJiNTI0MzBjOTAxMGVkYmUzOWZkYjg1ODg3YjdhZCxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) - the time in seconds since midnight, January 1, 1970. 

Computer Scientists thought this would be a good time to start measuring the time from because the first app was born at midnight, January 1, 1970. The app was a custom emoji pack for an ancient model of phone that would one day evolve to become the first Blackberry.

If you’re wondering why the response starts with “for (;;);”, it’s to, among other things, [encourage developers to use a quality JSON decode](https://t.umblr.com/redirect?z=http%3A%2F%2Fstackoverflow.com%2Fa%2F7099820&t=ZjcxZjBkMmVkNzMwODJjMTI3MWI0MzczMzIwM2FhNjk0NWExMTExYixZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0)r, instead of like, y’know, eval().

Anyway that “la” thing stands for “last active”, and tells you the last time the person was active on [Facebook](https://t.umblr.com/redirect?z=http%3A%2F%2Fnsa.gov%2Fcareers&t=MWNlZTQ5YzdkNzY0NzAwZTYzOTYyODg1YmFjNDQzZWVlNzc3NzY4OCxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0), down to the _second_. Do you see where I’m going with this?

Roleplaying as the NSA ˙ ͜ʟ˙
----------------------------

So far we have a whole bunch of things which look like this

*   A person
*   A time
*   Whether they’re **online** or **offline** or **idle**
*   Which devices they’re online/offline/idle on  
    

This doesn’t seem that interesting at first, since you already know who is online by looking at the sidebar. But **what if there was someone _always_ watching the little green dots?**

Using the _power of computers_, you can just write a Python program to listen to what the /pull requests are saying _all the time ever,_ and write it down.

Here’s a screenshot of all the log files I’ve got:

![](https://media.tumblr.com/8baab8b51d210006908dcbcd22c54f5a/tumblr_inline_o2l7jvJlHs1t2msv6_500.png)

And here’s what an individual log file looks like (the first 10 lines):

![](https://media.tumblr.com/cbfd4989fdf0ec3f13d9d3fd856daeb1/tumblr_inline_o2l7icOn6R1t2msv6_500.png)

Those blurred out things are [Facebook](https://t.umblr.com/redirect?z=https%3A%2F%2Ffacebook.com%2F&t=OWRjMDdkOWQ4NDBhNWEzZDM0YjY0NWVhMzlkYjVhYTA0NzdmYjI5ZCxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) user ids. If you think these screenshots look _a little bit creepy_ then YEAH I KNOW RIGHT. 

Tell me about your program then you massive nerd
------------------------------------------------

It runs 24/7, and it’s constantly logging online/offline activity data from those /pull URLs using my [Facebook](https://t.umblr.com/redirect?z=http%3A%2F%2Fhtml9responsiveboilerstrapjs.com%2F&t=NWM2ZmE1MDg3MzMyNzdkZGMyYTZkZjFhMDUxZjcwZmIwYmMzN2U3MSxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) cookie.

Writing it was _mostly_ about saying “jeez, all these parameters look _complicated_” and then blindly copy/pasting them anyway.

![](https://media.tumblr.com/27fdf3e14ddc7a755a8cc69c85f1cd62/tumblr_inline_o2goqdY2zK1t2msv6_500.png)

Protip, you can right click on any network request in Chrome’s Developer Tools and click “Copy as cURL”. This is _amazing_ and lets you re-run a request from the terminal, as well as give you all the headers and cookies used to run that request in a nice copy-pasteable format.

The first step was to just run that request verbatim in a terminal with curl. 

> curl ‘[https://1-edge-chat.facebook.com/pull?channel=p_\[redacted\]&seq=3&partition=-2&clientid=\[redacted\]&cb=6dcn&idle=5&qp=y&cap=8&tur=1545&qpmade=1455427171900&pws=fresh&isq=221841&msgs\_recv=3&uid=\[redacted\]&viewer\_uid=\[redacted\]&sticky\_token=239&sticky\_pool=atn2c06_chat-proxy&state=active](https://1-edge-chat.facebook.com/pull?channel=p_[redacted]&seq=3&partition=-2&clientid=[redacted]&cb=6dcn&idle=5&qp=y&cap=8&tur=1545&qpmade=1455427171900&pws=fresh&isq=221841&msgs_recv=3&uid=[redacted]&viewer_uid=[redacted]&sticky_token=239&sticky_pool=atn2c06_chat-proxy&state=active "Linkify Plus Plus")’ -H 'origin: [https://www.facebook.com](https://t.umblr.com/redirect?z=https%3A%2F%2Fwww.facebook.com&t=MGEzYjliNjMyYjkzYTNkZDhmMTgwMWI1OWJmZTA2MDFmNmZjNTkyZixZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0)’ -H 'dnt: 1’ -H 'accept-encoding: gzip, deflate, sdch’ -H 'accept-language: en-US,en;q=0.8,en-AU;q=0.6’ -H 'user-agent: ‘\[redacted\]’ -H 'accept: */*’ -H 'referer: [https://www.facebook.com/’](https://t.umblr.com/redirect?z=https%3A%2F%2Fwww.facebook.com%2F%27&t=OTY1NjRkMDQ4ZjE0YWUxZmZhNDlkMzdkYzA5NDVkNGUzNDQxZTA4MixZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) -H 'authority: [1-edge-chat.facebook.com](http://1-edge-chat.facebook.com/ "Linkify Plus Plus")’ -H 'cookie: ‘\[redacted\]’ –compressed

I was expecting it to not work because it looks like it has some sequence numbers in it oh boy BUT it turned out to just take a really long time. I later found out this was because the /pull endpoint is using [HTTP Long Polling](https://t.umblr.com/redirect?z=https%3A%2F%2Fen.wikipedia.org%2Fwiki%2FPush_technology&t=YmY1NWEwZjM1NTFkZmFlYTU2YjRhNmIxZjU4Y2U4OGNkYzliNzU2NixZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0), which turns out to be like a streaming HTTP GET request.

The only other important parameter to worry about is “seq”, which I’m guessing is the sequence number of the response from Facebook. Just add 1 to the sequence number that the response from /pull gives for the next request and you’re good to go.

If you’re worrying about remembering all this, chill out I got yo’ back, my 100% [Terms of Service Compliant](https://t.umblr.com/redirect?z=https%3A%2F%2Fwww.facebook.com%2Fterms&t=ZTc4YzI4OTcwYjI5MDFjMDBkMDAzYTBmNmU3NTFjMTg5ZjYxMTc3ZSxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) implementation of this is available [here on GitHub](https://t.umblr.com/redirect?z=https%3A%2F%2Fgithub.com%2Fdefaultnamehere%2Fzzzzz&t=MjMwNThkMmU5YzQ0OTM5MWU3ZjY0NjVjZjNjZWIwYmFkNGFjYzI0NyxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0). Standard disclaimers of “I’m so sorry I wrote parts of this in like 30 minutes” apply.

One caveat of the data-collection program that I’ve noticed is that it has false negatives. That is, sometimes it won’t give you a “this person is online” data point, even though they really are online. I guess that gives plausible deniability of… being offline? 

You should probably get out more
--------------------------------

\[worried laughter\]

So that’s the hard part done, right?
------------------------------------

**Let me paint you a word-picture**. It’s 11pm, I’m listening to [the soundtrack to The Social Network](https://t.umblr.com/redirect?z=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3D1Ua2gabdJoc&t=ZWNmYjVlNWYxMjZhZmE2YjE0YjFmYzUyYzk1MDA5ZWFjNmYwMWQ3MSxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) (ironically? meta-ironically? I don’t even know), I have six terminals tiled across two screens as well as fifty thousand browser tabs open and I’m up to my _**third** graphing library_.

Making graphs is really hard.

I used [matplotlib](https://t.umblr.com/redirect?z=http%3A%2F%2Fmatplotlib.org%2F&t=Y2E3NzI2MzdlYTdjMmMxYTFkYjViMDQ4YWNmYTU4N2Y0ZDc2YjdlMCxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0), but I realised this wasn’t my thesis and I wouldn’t be embedding this ugly graph as a pdf into a LaTeX document that takes 3 passes of pdflatex to render because there’s been a terrible but extremely localised accident where only humanity’s LaTeX to pdf converters have been irreversibly sent back in time to the 80s.

I used [bokeh](https://t.umblr.com/redirect?z=http%3A%2F%2Fbokeh.pydata.org%2F&t=OTQyYzc4NDNkNmEzYzYwMjNjNTVkNDk4OTRiOWY2NWIxZjU3MDYzZSxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0), which claims to be a “matplotlib-killer”, and it was was okay until [a friend](https://t.umblr.com/redirect?z=http%3A%2F%2Fblakeapproves.com&t=NmMwZGVjNjI3MTJlYjRjNjlhMDMyNzk4ZWU3ZDA0NjhmMDBkZGU5NixZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) told me “it isn’t the 90s anymore, you don’t generate graphs server-side. Also your graphs are ugly and you should feel ugly **you utter fraud**”.

[This friend](https://t.umblr.com/redirect?z=http%3A%2F%2Fblaker.space&t=ODFkNTAxNTNkZGIxNjE3ODFjYTMzYTgxM2IxM2EyMGMxNzZlNzcwYixZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) recommended [nvd3.js](https://t.umblr.com/redirect?z=http%3A%2F%2Fnvd3.org%2F&t=M2JhOWNkMWU2ZjViYmEzZjA4NGQzNDk3ZmQxNDVjOTYyOTQ3NGIyYSxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0), presumably because you’re not making _real_ graphs in 2016 unless your graphing library is <something>.js and requires at LEAST one other <something else>.js as a dependency. Everyone looks at you like “**what, you DON’T already use <something else>.js?** Jeez say goodbye to your Hacker News karma. Just apt-get install npm && npm install bower && bower install-” NO STOP IT THIS ISN’T WHAT [TIM BERNERS-LEE](https://t.umblr.com/redirect?z=https%3A%2F%2Fen.wikipedia.org%2Fwiki%2FTim_Berners-Lee&t=NjNhNzYxMjc1YmQ4NTVjMDI4YjRhZWYwNmZlOGY5MmUwMTk3NTNmNSxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) WANTED”.

I think it took about three times as much time to graph the data as it took to write the code to download it. And **the graphs aren’t even good**! I gave up on perfecting the graphs so I could just hurry up and write this questionable blog post already. Just think of me resolving pip3 dependencies when you see the ugly graphs.

(°ロ°)☝ **AND ANOTHER THING** when it’s midnight and your x-axis formatting function doesn’t convert UNIX times into JavaScript date objects properly because there’s no timezone information and I dunno JavaScript was written by some guy in two weeks (yeah I ain’t afraid to call it out what of it) and your binary-search based conversion of sparse timeseries data into uniformly dense timeseries data is causing _so many data points_ to be graphed that it’s slowly crashing Chrome and you’re watching _helplessly_ as your RAM goes up and Chrome won’t close the tab and it just _doesn’t seem right_ that 2016, the year of the Linux Desktop has brought us this situation I mean I thought if you had enough <something>.js libraries this stuff was meant to just _scale right up_ so tha-

Quit stalling with graphing libraries and show me the graphs
------------------------------------------------------------

Fine but you’re missing out on top-quality graphing-related banter.

* * *

_The graphs in this section are all of the online/offline activity of some of my Facebook friends.They consented to it being on this blog post on the condition that it’s anonymous. _

* * *

Person 1
--------

![](https://media.tumblr.com/9136dd4e51085e103a872c5252c5a51a/tumblr_inline_o2gpwd7WUG1t2msv6_500.png)

Here’s someone’s graph. The x-axis is time, and the y-axis is how online the user is. Possible states for someone’s status are “offline”, “invisible”, “idle”, and “active”. Each coloured line is a different kind of _client._ It’s called a _client_ because I don’t know I’m an Information Visualisation _Professional_ and I get to make up words like that. Here are explanations for what each of the “_coloured lines”_ mean_s_

*   **status** \- Not sure what this is. Some kind of client-agnostic status? It doesn’t line up exactly with the activity of the other clients though  
    
*   **webStatus** \- Chat activity on [facebook.com](http://facebook.com/ "Linkify Plus Plus")
*   **messengerStatus** \- Status on the Messenger mobile app
*   **fbAppStatus** \- Status on the Facebook mobile app
*   **otherStatus** \- Presumably shows when people are online on other apps that can access the API that causes them to be considered “online”. OAuth? Random “apps” like Farmville? No idea

Here’s the same graph, with some clumsy drawings on it showing when I think this person is awake/asleep.

![](https://media.tumblr.com/1de7e3691413a7ff657d151193adc508/tumblr_inline_o2gpz7xQEU1t2msv6_500.png)

You can see the amount of rest they’re getting each day - it’s the width of the “asleep” bit.

You can also see that they were probably asleep from 3am to 10am on February 11, and BOY does it feel creepy writing this.

Of course, this isn’t perfect, since they might be awake and _not_ using Facebook (I know). Having spoken to a few people who were graphed, it’s been a fairly accurate measure of awake/asleep time, as well as “how much do you browse Facebook at work” time ;)  

Do you look at Facebook shortly after you wake up? Shortly before you sleep? If so, these graphs are a fairly accurate way to measure when you were asleep, and anyone you’re friends with on Facebook can do it.

Person 2
--------

![](https://media.tumblr.com/cce74c4f87c2995c8e30e740003d0f88/tumblr_inline_o2gqevRU6U1t2msv6_500.png)

I showed this person their graph and asked them some questions.

> “Did you go to sleep around 11:10pm last night?”  
> They said yes.
> 
> “Did you wake up around 8:32? That’s a weird time. Was your alarm set for 8:30?”  
> They said yes.

NSA APPROVED   ✔️ 🆗👌👌 👍✔️👌🆗🆗👍

Person 3
--------

![](https://media.tumblr.com/90d58c03732cc720bc1e7040f7321fee/tumblr_inline_o2gqzp9pBy1t2msv6_500.png)

There are two interesting things about this graph

*   The person isn’t online as frequently as the previous examples  
    
*   The person isn’t using the Messenger app nearly as much  
    

You can see that their webStatus was “online” on and off from midnight til around 2am, and then again at 10:21am.. I’m not sure if this spiky pattern means that they really were online, then offline, then online again, or if it’s just a quirk of the dodgy undocumented “API” I’m using, or even if it’s just a problem with my code. 

Similarly, I’m not sure why there are these weird spikes every three minutes (+- ~1minute) sometimes.

![](https://media.tumblr.com/d13cf0b4b09e07fc82b9f2a0af6faf41/tumblr_inline_o2gr56osi31t2msv6_540.png)

Also, why does “otherStatus” go to offline precisely when “webStatus” goes to online? So many questions! Let me know if you know the answers to any of these things (@Facebook employee friends ;) ;) ;))

* * *

Anyway, I hope I’ve convinced you that this is real creepy. I don’t really want to be able to have the power to do this.

Your dumb graph screenshots are too small. Give me a live graph to play with
----------------------------------------------------------------------------

[You got it, boss. Click here. Or anywhere, really. This whole sentence is a link.](https://t.umblr.com/redirect?z=https%3A%2F%2Fcdn.rawgit.com%2Fdefaultnamehere%2Fzzzzz%2F70d407736092304ee247fbbacbe9f82bc0cba472%2Ftemplates%2Fsample_graph.html&t=MjU0Nzg4NDYwZGViZDE1N2RjZjVmNzg1YzU3MTJkY2I3Y2IwYzU3ZSxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) 

What else can you do with this data?
------------------------------------

You can aggregate. Finding the average wake up time/sleep time/time spent on Facebook each day and then looking for outliers sure sounds like a way to find interesting things about your Facebook friends.

You can write a thing to email you every morning with the names and sleep times of everyone who’s had less than 6 hours of sleep.

You could even try and guess when your friends are talking to each other, by looking for times when only a few people are active, although I suspect this would be hard.

I’m sure you can come up with something else, too.

Why can you do this? Can’t Facebook stop this from happening?
-------------------------------------------------------------

That’s a good question, thanks for asking.

It makes sense for [Facebook](https://t.umblr.com/redirect?z=http%3A%2F%2Fnicememe.website&t=ZTExMGM3NTMxYjA1MjA0MjUxMzY4MGRjZmUzZGM0Mzk2NTc2ZTM3ZCxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) to be able to do this, since they can tell when everyone is online anyway. But why can your [Facebook](https://t.umblr.com/redirect?z=https%3A%2F%2Fmatmartinez.net%2Fnsfw%2F&t=YjFiN2FmZTkyMzZhZmQ5ZjMwMzljMzQzY2M4NTFkYzc3YzkwNjQzNyxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) friends do this to you? 

I don’t know all the details of how [facebook.com](http://facebook.com/ "Linkify Plus Plus") uses all the data that’s sent via the /pull endpoint, but it’s kinda creepy that I can see my friends’ status on every device? I guess they could just give me “web” or “mobile” or “offline”, rather than the full list of statuses for every client, but even that doesn’t solve the problem.

I also see the value in seeing “last active 4h ago” and “last active 1m ago” for Messenger contacts but… I dunno, here I am making these creepy graphs. 

Anyway, I just open-sourced my dodgy graph making thing so now everyone can do this. And who knows how many people have been doing it already?

I’m probably oversimplifying it, though. The smart people at [Facebook](https://t.umblr.com/redirect?z=https%3A%2F%2Fmyspace.com%2F&t=ZDZkOWQzMDlmYTE1MzA0YzdkMDhiNjM0OThhN2RmNWM3YWIyYzNiMixZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) who write this stuff have probably thought of all of this and found that this way was best.

Can I stop you from doing this to me?
-------------------------------------

Kinda. Coincidentally, because my script is always running, collecting data, I show up as “online” all the time. If you were also running a script like this, it would partially prevent what I’m doing from working on you, since you always show up as “online”, no matter what you’re _really_ doing. Activity from the Messenger app will still show up separately, though.

tl;dr
-----

*   [Facebook](https://t.umblr.com/redirect?z=https%3A%2F%2Fpbs.twimg.com%2Fmedia%2FCRRJVwIUwAAf-wP.png&t=NTU0ODEzZDgyMzVmMzc0MzdjMjgwMjNiNTIwNzlkOWE5NGVmZGFlMSxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) sends your computer a bunch of interesting information when you’re on [facebook.com](https://t.umblr.com/redirect?z=https%3A%2F%2Fpbs.twimg.com%2Fmedia%2FCRRGTVcU8AA7veC.png%3Alarge&t=NDMzZWUwOGNkM2Q1YzlhMzU4NDYxNjA5NTg5MjQ4YzI3MGI2YzNhMixZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0).   
    
*   You can collect that information over time and use it to keep track of when people are on [Facebook](https://t.umblr.com/redirect?z=http%3A%2F%2Fdudududu.de%2F&t=NWNhYmUxMjk1Y2NjMjRmZWQ4ZDkxM2ZlYmJiNjk1OGU3NjAzOTIwMyxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0), and which devices they’re using.  
    
*   You can make a pretty good guess as to what time people are going to sleep and waking up
*   It’s creepy, but I don’t see a way for [Facebook](https://t.umblr.com/redirect?z=https%3A%2F%2Fen.wikipedia.org%2Fwiki%2FPhishing%23Link_manipulation&t=OWY0MTFlNzMzYzBjZTBkZDc4M2Q2NDA3ZTE0MWRhMDVhNTJkMzJkZixZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) to stop allowing this while still making their chat app good.

So how does this make money again?
----------------------------------

Oh, no no no. I just uh don’t get out much.

* * *

_If you want to talk to me about this blog post then I dunno [tweet at me](https://twitter.com/_notlikethis) I guess. You can also [stalk me on GitHub](https://t.umblr.com/redirect?z=https%3A%2F%2Fgithub.com%2Fdefaultnamehere%2F&t=ODU5YzNhMzViNmVlNjllNWUyN2E0OWI5M2ZkZmZjMDc2MTQxYjNkOSxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0) if you want._

_For the latest in dumb novelty websites, please direct your browser to [https://mango.pdf.zone](https://t.umblr.com/redirect?z=https%3A%2F%2Fmango.pdf.zone&t=NmE5Yjg4ZjIwMzNiOGQ5OTY3ODVmMTA5MDdjYTQ2YzYwYzQ1Y2RlMCxZRkpQNkM5SQ%3D%3D&b=t%3AjB2RAmXOqO7xFf4uOeN6Fg&p=https%3A%2F%2Fdefaultnamehere.tumblr.com%2Fpost%2F139351766005%2Fgraphing-when-your-facebook-friends-are-awake&m=0)_
